### ISSUE TITLE
[Wallet] Sending Stratis times out

### ISSUE THREAD

**Issue by dev0tion** [stratisproject/StratisBitcoinFullNode#597]
As reported by @mtaipe:
"Click send BITCOIN SENDs and confirms message shows  also STRATIS  sends but no confirmation and modal doesn’t close"

This happens every now and then, sometimes it does get the confirmation window, sometimes it doesn't. To me it looks like the build-transaction is working, but the send-transaction call sometimes times out, thereby not triggering the confirmation window on the UI.

**Comment by dev0tion** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [[Wallet] Sending STRAT times out](https://trello.com/c/FSz0ebHu/294-wallet-sending-strat-times-out)

**Comment by nopara73** [comment]
Debugged it on Friday.   
Sendtransaction works as intended, it's just if nobody propagates back the message to our node, then we cannot be sure the message was actually propagated. Therefore it times out and the transaction comes when it gets confirmed in a block.

**Comment by dev0tion** [comment]
Ok, but is there any way I can catch this in the UI?

**Comment by nopara73** [comment]
Yes, you got a bad request after 21 seconds of tx not arriving.  

`while (TimeSpan.FromSeconds(21) > waited)`  

`return this.StatusCode((int)HttpStatusCode.BadRequest);`

**Comment by bokobza** [comment]
@nopara73 @dangershony A while ago we talked about sending the trx to the network and then returning to the called straight away (hence relying on the network to propagate it back to us).
I guess that didn't make the cut. 
Can we not reconsider it? Maybe the UI can wait for the trx to show up and show a message if nothing shows up after 21 seconds?

**Comment by nopara73** [comment]
>  A while ago we talked about sending the trx to the network and then returning to the called straight away (hence relying on the network to propagate it back to us).  

What do you mean by this? I don't understand.

> Maybe the UI can wait for the trx to show up and show a message if nothing shows up after 21 seconds?  

That's how it is now, but without the message after 21 seconds.

**Comment by dangershony** [comment]
We had a discussion here and decided a good approach.
We add the trx to the wallet and mark it with a flag (PendingPropagation) in the UI you can grey it out (or set a flag).
Once propagation is confirmed it will be unconfirmed but propagated (as is now). 

We will then provide the user the ability to delete propagated trx (right click and delete)

**Comment by nopara73** [comment]
> We will then provide the user the ability to delete propagated trx (right click and delete)

If I remember well, we discussed that we don't want to delegate such confusing burden to the user. @bokobza

**Comment by dangershony** [comment]
Small correction to my last comment 
>We will then provide the user the ability to delete "unconfirmed" and "unpropagated" trx

**Comment by nopara73** [comment]
@dangershony  How long did it take you to actually understand the concept of unconfirmed transactions? It took me half a year. If I would've start Bitcoin with Breeze, it'd take me half a year to confidently use that feature.  
Now think about Breeze's target audience: your grandma. How long would it take her to confidently use this feature?

**Comment by dangershony** [comment]
Yeah the `unpropagated` was just a way to call that scenario I definitely wont put such text on breeze.
As far as normal users are concerned its `unconfirmed`.

**Comment by noescape00** [comment]
Suggestions and thoughts about how to refactor code related to that issue: 

0) Add an additional flag that says the trx was sent but no inv messages came back so we wont time out we broadcast and set a flag and when we hear back form peers we flip the flag. 

1) rename * BroadcasterManager * => * BroadcastManager *
 in all classes. 

2) rewrite all that "oneTwo = oneTwo == 1 ? 2 : 1;" stuff. (FullNodeBroadcasterManager:62, LightWalletBroadcasterManager:48)

3) "ask half of the peers if they're interested in our transaction" why half? maybe we should ask <= TargetPeersTxBroadcastCount (add this constant)

4) "..foreach (var var node in this.connectionManager.ConnectedNodes)" - it's not a node. Rename to networkPeer? (FullNodeBroadcasterManager:54,  LightWalletBroadcasterManager:42)

5) "await node.SendMessageAsync(invPayload).ConfigureAwait(false);"  why do we need to broadcast it one by one, maybe do it in parallel- all at once?

6) looks like SmartBitBroadcasterManager is not used. Do we actually need that? It looks like it depends in 3d party api (smartbit.com.au/api). I suppose it can be used in case we don't have any connections and we want to propagate tx as fast as possible. But in that case we need some kind of selector that would automatically select this broadcaster in case we don't have any connections or connection count is insufficient. Though in my opinion btc network is huge enough- I always get like 5-10 connections a few seconds after the node launched. And we can't use this one for stratis. So we probably don't even need this 3rd party api dependency at all. 

7) xmldoc for everything

8)rename  InvPayload => InventoryPayload

9) InvPayload: replace private\public inventory with a field that has private set and public get. 

10) InvPayload: remove redundant functions, remove inheritage of IBitcoinSerializable because it already inherited from Payload (which implements IBitcoinSerializable)

11) TxFeeEstimateRequest.Amount- change it's type from string to Money.

**Comment by noescape00** [comment]
>The purpose of the light node is that we don't save the blocks we download, we merely parse them to see if they contain transactions affecting our wallets.
>Downloading the whole mempool is not a good idea and something we definitely don't want to do.

Ok. So we can't download the mempool and we can't ask peers for a single transaction because it will reveal us as the owners of transaction. 

Do we have any ideas on how to effectively improve current algo? 

Copy-paste from PR 844, description of current algo: 



**Current light node algorithm for verifying the transaction and knowing if it was propagated:**

1.  Create a transaction.
2.  Broadcast it to half of the peers that we are connected to.
3.  Wait until another half of the peers propagate it back to us.
4.  If we see out tx being propagated from any peer back to us we know that tx is valid and propagated.

**Problems of this algorithm:** 

- If we don't get transaction propagated back to us we end up with timeout exception.
- We will always get a timeout exception if we are only have 1 peer connected.
- Since we use only 50% of the peers the transaction propagates slower than it potentially can.
- If we successfully propagated tx but peers that should propagate it back to us are disconnected and no new peers got connected within a time frame of 21 second we end up with timeout exception.

**Comment by Aprogiena** [comment]
some conversation happened in #851

**Comment by noescape00** [comment]
If there are no suggestions on how to effectively replace the old algorithm then we probably can improve it? 

Steps we can take: 

- Increase timeout from 21 sec to >21 sec. 
- Allow sending transactions only when we have at least 2 peers so the algo can work correctly. 

But I don't feel like those are correct steps. 

Also we can stop using this algo at all and get info that our tx is valid only when we receive a block that contains it. 

Or we can still implement the new algo but instead of asking just 1 tx that we are interested in we can also ask for random transactions as well (probably for ones that were propagated earlier to us). But that wont work if even full nodes have 1-2 mempooled txes on average at this stage of stratis. It could be a good change for the time when stratis is a popular enough network with lots of txes going around.

**Comment by noescape00** [comment]
By the way why light node is not just a pruned full node? 
If we change what light node is we won't be facing this issue.

**Comment by dangershony** [comment]
Pruned node takes still several GB in disk.

This is what I suggest in terms of algorithm:

- We keep a `broadcastpool` of trx that are send from our node and not included in a block yet.
- We keep a flag on a trx that indicates if it was propagated or not, the result is `IsPropagated=true` means the trx its now in an unconfirmed state until a block arrives that has the trx in which case its removed from the `proadcastpool`, the flag does not necessarily need to live on the trx itself but it could be a value set in the proadcastpool item, trx that are not marked as broadcast after some time can be tried again, if we keep the flag in the trx itself then a node restart one break the processes.
- The `broadcastpool` has a background loop that broadcasts a trx to peers every X seconds (say every 10 seconds) with fixed delays, say when a trx just arrives its pushed to one peer, 10 seconds later to another and so on until all peers are marked a sent, a peer that request a getdata of that trx is considered broadcasted, if we receive an inv form a peer the trx was not send to then its considered `IsPropagated=true` and we might decide to stop sending it on (or send in bulk to the rest of the peers).

This approach can work for both fullnode and lightnode with the exception that the fullnode will first include the trx in the mempool before broadcasting it out. 
 
Thoughts?

**Comment by noescape00** [comment]
> if we receive an inv form a peer the trx was not send to then its considered IsPropagated=true and we might decide to stop sending it on

I'm not sure how is it different from the current approach. We still rely on peers that we did not send the tx to to propagate it to us to consider it valid and propagated. So this approach will have the same problems as current one does. 

But the idea of broadcast pool is good. It solves the problem of storing transactions that are not propagated (or we don't know if they are).

**Comment by noescape00** [comment]
>propagation will not happen with half of the nodes but incrementally
That will slow the propagation though. 

Why don't we use a mix: send it to half of the nodes and then send it one by one to the rest ones every n seconds.

**Comment by noescape00** [comment]
But in general I think yeah, that solves that timeout exception and gives an additional API for UI wallets to know the state of the transaction. 

So basically we would return success on SendTransactionAsync right away and after that for both full nodes and light nodes we would need to ask that broadcastpool if our tx is propagated in case we need to know that.

**Comment by noescape00** [comment]
So I guess we've come to a solution on how to implement this. 

@bokobza @Aprogiena do you have any thoughts on this, are you ok with me doing this as it was suggested in this issue?

**Comment by bokobza** [comment]
Wait, I want to read all of this first :-)

**Comment by Aprogiena** [comment]
I haven't followed recently, I'm quite overwhelmed with other stuff. I have mentioned the privacy concerns, so as long as the new solutions is not worse for privacy I'm ok with that, otherwise I'd be quite strongly against.

**Comment by bokobza** [comment]
Let's wind back for a minute and think about what we're trying to achieve.
We want our users to **send funds** and know whether the operation was **successful or not**.
Now we know a couple of things:
* ‎hundreds of transactions have been sent with the current code
 * ‎transactions have trouble propagating when there is a low number of peers, but they sometimes do
 * we occasionally receive a timeout but sometimes the transaction has actually been sent successfully

With that in mind, trying to"fix" transaction sending is like trying to fix loading a website when internet is flaky. It's just gonna be too complicated, there's gonna be too many hoops to jump through and it's hard to maintain and test.

When I pay for something online I'm happy to wait to make sure the transaction has gone through.
I also wouldn't want to use a wallet that leaves me, the user, in limbo. The user needs to know right now whether his transaction has been sent or not, and not "we'll retry for you in a bit and we'll do that until we succeed".  With this approach the user is uncertain of what's happening and he also can't cancel this transaction. If he can, that's yet another level of complexity that needs to be added. 
The user should be able to try, maybe try again if it didn't work, and then decide to stop or continue.

Timeouts are a thing that we can all live with. We just need to mitigate failures and communicate it to the user clearly.

Looking at the list of things we know above, it seems to me that in a first instance we should raise the timeout to 25 seconds and only propagate when we have at least 2 peers.

**Comment by noescape00** [comment]
Hmm. I can see the logic behind that. 

>and only propagate when we have at least 2 peers

We can do that, indeed. Especially taking in count that @fassadlr will soon improve the peer selection so it's easier to get at least 2 peers almost right away. 
But the overall requirement of having 2 connections to send tx... Not sure if I'm a fan of that.

**Comment by noescape00** [comment]
So there are 2 approaches now: 
1) @bokobza's make requirement of 2 peers and increase timeout
2) @dangershony's : make a pool

@dangershony what do you think about @bokobza's comment? 

When the agreement is reached I'll start working on the PR.

**Comment by dangershony** [comment]
>I also wouldn't want to use a wallet that leaves me, the user, in limbo

If you get a timeout you are already in limbo world as we don't know what really happened.
Is it not better to keep some sort of flag to track propagation?

>When I pay for something online I'm happy to wait to make sure the transaction has gone through.

What happens if when you pay for something online you get a timeout and you don know if you paid or not?

**Comment by dangershony** [comment]
By the way 25 seconds is a long time for propagation I don't think its that long in reality.
I remember reading some where that it can take up to 12 seconds to propagate globally to the entire network, so increasing the timeout wont help really.

**Comment by noescape00** [comment]
>By the way 25 seconds is a long time for propagation I don't think its that long in reality.
I remember reading some where that it can take up to 12 seconds to propagate globally to the entire network, so increasing the timeout wont help really.

I think such long timeout is here just in case we disconnect from the peer we waiting to propagate it back. So 25 sec includes time we need to get some more peers, not for the actual propagation.

**Comment by bokobza** [comment]
So @dangershony and I had a chat and we're thinking that the send transaction **could** be fire and forget: no pool, no retry, no wait, no timeout. So we send the transaction to the network and we add it to the wallet straight away as unconfirmed.

With this approach, let's break down the scenarios and zero-in on what problems still need solving:

If sending a transaction is successful straight away:
- trx goes in the wallet as unconfirmed 
- confirmed when put in a block

If sending a transaction takes time to propagate:
- trx goes in the wallet as unconfirmed 
- confirmed when put in a block (maybe a bit later, maybe in the next block)

If transaction is invalid and will never be confirmed:
- trx goes in the wallet as unconfirmed 
- stays there forever

Clearly, the only scenario we need to find a solution for is the last one. If a trx is invalid and will never be confirmed, how do we know? and what do we do? What if the user sends a lot of transactions after the first one, reusing the same outputs? All his transactions will be void and he'd be in for quite a surprise when trxs never confirm.

A way to go would be to have a flag "pending" inside the transaction object in the wallet. When we send a transaction and we save it in the wallet, we set "pending" to true, and when we receive this transaction via a peer or in a block, we remove the "pending" field. 
Having a flag will allow us to easily spot old/invalid/dead transactions in the wallet, in order to show it in the UI/allow users to remove them/be smarter about which outputs to choose when sending subsequent trxs.

We can also show an error if the user tries to send with only 1 peer, that will make these event even more rare.

@noescape00 @dev0tion  what do you think?

**Comment by bokobza** [comment]
Actually what I proposed here isn't great either.
We may not need a flag but instead just look at the fact that the transaction is still unconfirmed and old.
We can then offer the user the ability to delete it from the wallet after a set time (like an hour or something).

**Comment by noescape00** [comment]
So basically what I'm doing is: 

1) Set min 2 nodes requirement to send tx.
2) Save tx to the wallet with a state=unconfirmed. When new block is received check if it contains our tx and if it does change state to =confirmed.
3) If the peers propagate tx back set state to: pending. 
4) We DO NOT rebroadcast the transaction after it was broadcasted once it was created. 

Ok, sounds good. Not the optimal solution overall but looks like there is no better.

**Comment by Aprogiena** [comment]
I don't think step 1 is a good thing. This would disqualify custom setups (among other cases) where you use `-connect` to connect to a single specific node. This is perfectly valid scenario. If we forbid transactions to be propagated in such scenario, imho that's bad decision that will remove certain use cases.

**Comment by bokobza** [comment]
1. @Aprogiena is right, I hadn't thought of that. Minimum 1 node then. Maybe in sendTransaction, check there is at least 1 node connected, otherwise show a friendly error.
2. You don't need to do that, it's already done. When the node receives a transaction, it puts it in the wallet and it doesn't have a blockHeight. When a block containing this transaction is received, the transaction is updated to include the blockHeight. This is how we know unconfirmed (no blockHeight) vs confirmed (blockHeight is set).
3. mnaybe add a Propagated flag. So when we send the transaction we insert it as Propagated=false and when we receive is back we set it to true.
4. correct
